feedback/ngffeedback.cpp

```diff
@@ -0,0 +1,318 @@
+                    m_client.stop(id);
+            }
+            qCDebug(ngflc) << "Stopped all effects";
+        }
+    }
+}
+
+QVariant NGFFeedback::actuatorProperty(const QFeedbackActuator &, ActuatorProperty prop)
+{
+    switch (prop) {
+    case Name:
+        return QLatin1String("NGFD");
+    case State:
+        return QFeedbackActuator::Ready;
+    case Enabled:
+        return m_actuatorEnabled;
+    default:
+        return QVariant();
+    }
+}
+
+bool NGFFeedback::isActuatorCapabilitySupported(const QFeedbackActuator &, QFeedbackActuator::Capability)
+{
+    /* Changing envelope or periodicity is not supported
+     * since we don't support changing intensity level.
+     */
+    return false;
+}
+
+void NGFFeedback::updateEffectProperty(const QFeedbackHapticsEffect *effect, QFeedbackHapticsInterface::EffectProperty prop)
+{
+    if (!m_actuatorEnabled)
+        return;
+
+    if (!m_customEffects.contains(effect))
+        return;
+
+    if (prop == QFeedbackHapticsInterface::Duration) {
+        qCDebug(ngflc) << "Playing custom effect due to property update (" << effect->duration() << "ms)";
+        setEffectState(effect, QFeedbackEffect::Running);
+    }
+}
+
+void NGFFeedback::setEffectState(const QFeedbackHapticsEffect *effect, QFeedbackEffect::State state)
+{
+    if (!m_actuatorEnabled)
+        return;
+
+    quint32 id;
+    switch (state) {
+    case QFeedbackEffect::Running:
+        id = m_customEffects.value(effect);
+        if (id && m_activeEffects.value(id).state == QFeedbackEffect::Paused)
+            resumeCustomEffect(effect);
+        else
+            startCustomEffect(effect);
+        break;
+    case QFeedbackEffect::Stopped:
+        stopCustomEffect(effect);
+        break;
+    case QFeedbackEffect::Paused:
+        pauseCustomEffect(effect);
+        break;
+    case QFeedbackEffect::Loading:
+    default:
+        // not supported
+        break;
+    }
+}
+
+QFeedbackEffect::State NGFFeedback::effectState(const QFeedbackHapticsEffect *effect)
+{
+    quint32 id = m_customEffects.value(effect);
+    if (id)
+        return m_activeEffects[id].state;
+    return QFeedbackEffect::Stopped;
+}
+
+void NGFFeedback::startCustomEffect(const QFeedbackHapticsEffect *effect)
+{
+    if (effect->duration() > 0) {
+        qCDebug(ngflc) << "Playing custom effect due to state change (" << effect->duration() << "ms)";
+        QMap<QString, QVariant> properties;
+        properties.insert(QStringLiteral("haptic.duration"),
+                          QVariant(static_cast<quint32>(effect->duration())));
+        /* The choice of the effect will only affect the strength and style
+         * of the feedback. Duration is determined by haptic.duration property
+         * which either repeats or "stretches" the effect to the whole duration.
+         */
+        quint32 id = m_customEffects.take(effect);
+        if (id) { // Existing effect
+            m_client.stop(id); // TODO: Does this make any sense?
```

Comment:
Should I stop the effect so I can change the duration or can I just alter it somehow?

Reply:
Best to call stop for the event before starting a new one.